### PR TITLE
feat: lazy load and hot reload for policies

### LINKED ISSUE #870: Lazy load policies
Hi! I've been using Express Gateway on a project I am currently working on. It has been great to work with, thank you very much for the project!

I recently started implementing a custom policy for SAML authentication (using passport) and ran into a bit of an issue. It looks like when EG starts up it loads all of the policies that are included with it regardless if you are using them or not. It looks like the oauth2 policy adds a serializer and deserializer to passport when the module loads. Sadly, the deserializer was causing issues for me because Passport is invoking the oauth deserializer before mine and fails. 

I was wondering if there is a proper way to prevent modules from being loaded. If not, I was wondering if it would be worth it to add the ability to NOT load policies you're not using. Perhaps the loaded policies can be driven off of the `policies` array in the gateway's config file. 

I found a (hacky) workaround for now. Just wanted to check to make sure I am not missing anything.

Thanks!!!

### PR DESCRIPTION
**Pull request by ravikp7**
- only load policies defined in the gateway config

fixes #870

### CODE REVIEW

File: lib/gateway/index.js [hunk 1]
```diff
@@ -11,9 +11,9 @@ module.exports = function ({ plugins, config } = {}) {
   const appPromises = [];
   const apps = {};
   if (plugins && plugins.policies && plugins.policies.length) {
-    plugins.policies.forEach(p => {
-      log.debug('registering policy', p.name);
-      policies.register(p);
+    plugins.policies.forEach(policy => {
+      log.debug(`registering policy ${policy.name}`);
+      policies.register(policy);
```
Comment (XVincentX) [inline lib/gateway/index.js]: Good idea on the renaming.

File: test/e2e/oauth2-authorization-code.js [hunk 2]
```diff
@@ -35,7 +35,7 @@ describe('oauth2 authorization code grant type', () => {
           path: '/not-found'
         }
       },
-      policies: ['oauth2', 'proxy'],
+      policies: ['cors', 'jwt', 'basic-auth', 'oauth2', 'proxy'],
```
Comment (XVincentX) [inline test/e2e/oauth2-authorization-code.js]: These policies do not seem to be used in this test. Why did you need to load them?

File: test/e2e/policy-seq-oauth2-expression-log-ratelimit-proxy.js [hunk 3]
```diff
@@ -39,7 +39,7 @@ describe('E2E: oauth2, proxy, log, expression, rate-limit policies', () => {
           scopes: ['authorizedScope']
         }
       },
-      policies: ['oauth2', 'proxy', 'log', 'expression', 'rate-limit'],
+      policies: ['jwt', 'oauth2', 'proxy', 'log', 'expression', 'rate-limit'],
```
Comment (XVincentX) [inline test/e2e/policy-seq-oauth2-expression-log-ratelimit-proxy.js]: Same here, why did you need to add this policy?

File: test/e2e/oauth2-authorization-code.js (hunk 2 above)
Comment (ravikp7) [reply test/e2e/oauth2-authorization-code.js]: The tests were failing
Logs:
```
  606 passing (2m)
  2 pending
  2 failing

  1) oauth2 authorization code grant type
       "before all" hook:
     Error: Timeout of 60000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves. (/home/ravi/oss/express-gateway/test/e2e/oauth2-authorization-code.js)
  

  2) oauth2 authorization code grant type
       "after all" hook:
     TypeError: Cannot read property 'kill' of null
      at Context.done (test/e2e/oauth2-authorization-code.js:131:20)
```
With `LOG_LEVEL=debug`
```
  5) oauth2 authorization code grant type
       "before all" hook:
     Error: connect ECONNREFUSED 127.0.0.1:33635
      at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1097:14)

  6) oauth2 authorization code grant type
       "after all" hook:
     TypeError: Cannot read property 'kill' of null
      at Context.done (test/e2e/oauth2-authorization-co
```
Since, only policies loading was changed, so best guess was to try including some to fix those.
I thought that in the test in proxy `changeOrigin` is set to true, so `cors` need to be included, and others `jwt` because of `oauth2`. And no idea about `basic-auth`, but without it also, tests was breaking.

Comment (XVincentX) [reply test/e2e/oauth2-authorization-code.js]: Ok do not worry about these now, I can try to have a look into later.

File: test/e2e/policy-seq-oauth2-expression-log-ratelimit-proxy.js (hunk 3 above)
Comment (ravikp7) [reply test/e2e/policy-seq-oauth2-expression-log-ratelimit-proxy.js]: For this one, with debug
```
  605 passing (53s)
  2 pending
  2 failing

  1) E2E: oauth2, proxy, log, expression, rate-limit policies
       "before all" hook: setup:
     can't resolve reference jwt.json from id http://express-gateway.io/schemas/policies/oauth2.json#
  

  2) E2E: oauth2, proxy, log, expression, rate-limit policies
       "after all" hook: cleanup:
     TypeError: Cannot read property 'close' of undefined
      at Context.after (test/e2e/policy-seq-oauth2-expression-lo
```
So added `jwt` policy

File: lib/gateway/index.js [hunk 4]
```diff
@@ -49,11 +44,19 @@ module.exports = function ({ plugins, config } = {}) {
...
+const bootstrapPolicies = ({ app, plugins, config } = {}) => {
+  if (plugins && plugins.policies && plugins.policies.length) {
+    plugins.policies.forEach(policy => {
+      log.debug(`registering policy ${policy.name}`);
+      policies.register(policy);
+    });
+  }
+
+  // Load policies present in config
+  if (config.gatewayConfig.policies) {
```
Comment (XVincentX) [inline lib/gateway/index.js]: This field is required in the JSON Schema, so there's no chance it will be undefined. You can remove this check.

File: lib/gateway/index.js [hunk 5]
```diff
@@ -76,23 +79,37 @@ function bootstrap({ plugins, config } = {}) {
...
+        app,
+        plugins: pluginsLoader.load(newConfig),
+        config: newConfig
+      });
+      rootRouter = pipelines.bootstrap({ app: express.Router(), config: newConfig });
+      log.info('hot-reload config completed');
     } catch (err) {
-      log.error('Could not hot-reload gateway.config.yml. Configuration is invalid.', err);
+      log.error('Could not hot-reload gateway.config.yml. Configuration is invalid. ', err);
+      log.debug('Loading old configuration');
```
Comment (XVincentX) [inline lib/gateway/index.js]: Let's ditch this.

File: lib/gateway/index.js [hunk 6]
```diff
@@ -76,23 +79,37 @@ function bootstrap({ plugins, config } = {}) {
...
+      bootstrapPolicies({
+        app,
+        plugins: pluginsLoader.load(newConfig),
+        config: newConfig
+      });
+      rootRouter = pipelines.bootstrap({ app: express.Router(), config: newConfig });
+      log.info('hot-reload config completed');
     } catch (err) {
-      log.error('Could not hot-reload gateway.config.yml. Configuration is invalid.', err);
+      log.error('Could not hot-reload gateway.config.yml. Configuration is invalid. ', err);
```
Comment (XVincentX) [inline lib/gateway/index.js]:
```suggestion
      log.error(`Could not hot-reload gateway.config.yml. Configuration is invalid. ${err}`);
```

File: lib/gateway/index.js [hunk 7]
```diff
@@ -49,11 +44,19 @@ module.exports = function ({ plugins, config } = {}) {
     });
 };
 
-function bootstrap({ plugins, config } = {}) {
-  const app = express();
-  app.set('x-powered-by', false);
+const bootstrapPolicies = ({ app, plugins, config } = {}) => {
+  if (plugins && plugins.policies && plugins.policies.length) {
+    plugins.policies.forEach(policy => {
+      log.debug(`registering policy ${policy.name}`);
```
Comment (XVincentX) [inline lib/gateway/index.js]: Can we add in this statement that this is coming from a plugin?

File: lib/gateway/index.js [hunk 8]
```diff
@@ -49,11 +44,19 @@ module.exports = function ({ plugins, config } = {}) {
     });
 };
 
-function bootstrap({ plugins, config } = {}) {
-  const app = express();
-  app.set('x-powered-by', false);
+const bootstrapPolicies = ({ app, plugins, config } = {}) => {
```
Comment (XVincentX) [inline lib/gateway/index.js]: The `app` parameter does not seem to be used. Let's get rid of it.
```suggestion
const bootstrapPolicies = ({ plugins, config } = {}) => {
```

File: lib/gateway/index.js [hunk 9]
```diff
@@ -76,23 +79,37 @@ function bootstrap({ plugins, config } = {}) {
...
+};
+
+function bootstrap({ plugins, config } = {}) {
+  let rootRouter;
+  const app = express();
+  app.set('x-powered-by', false);
   app.use(passport.initialize());
+  bootstrapPolicies({ app, plugins, config });
   rootRouter = pipelines.bootstrap({ app: express.Router(), config });
   app.use((req, res, next) => {
```
Comment (XVincentX) [inline lib/gateway/index.js]: I think we can refactor this to be `app.use(rootRouter)`

File: lib/policies/index.js [hunk 10]
```diff
@@ -49,14 +49,25 @@ schemas.register('internal', 'baseAuth', {
...
-  const policy = require(path.resolve(__dirname, name));
-  policies[name] = Object.assign(policy, { name });
-  register(policy);
-});
+const load = (policiesToBeLoaded) => {
+  const policyNames = fs
+    .readdirSync(path.resolve(__dirname))
+    .filter(dir => fs.lstatSync(path.resolve(__dirname, dir)).isDirectory());
+
+  const policiesWithoutPlugins = policiesToBeLoaded.filter((name) => policyNames.includes(name));
```
Comment (XVincentX) [inline lib/policies/index.js]: Let's use a better name

```suggestion
  const corePolicies = policiesToBeLoaded.filter(name => policyNames.includes(name));
```

File: lib/policies/index.js [hunk 11]
```diff
@@ -49,14 +49,25 @@ schemas.register('internal', 'baseAuth', {
...
+
+  policiesWithoutPlugins.forEach((name) => {
```
Comment (XVincentX) [inline lib/policies/index.js]:
```suggestion
  corePolicies.forEach(name => {
```

File: lib/gateway/index.js [hunk 12]
```diff
@@ -76,23 +79,37 @@ function bootstrap({ plugins, config } = {}) {
...
 
   eventBus.on('hot-reload', (hotReloadContext) => {
+    const oldConfig = config;
+    const oldPlugins = plugins;
     const oldRootRouter = rootRouter;
     try {
-      rootRouter = pipelines.bootstrap({ app: express.Router(), config: hotReloadContext.config });
-      log.info('hot-reload router completed');
+      const newConfig = hotReloadContext.config;
+      policies.removeAll();
```
Comment (XVincentX) [inline lib/gateway/index.js]: I do not think this is a good idea. What we should be doing here is to check out and load only the *new* policies instead of ditching everything and reload them from scratch. It should not be a big deal, but I believe it's important.

Comment (ravikp7) [reply lib/gateway/index.js]: This was added to cater the case when a policy is removed from the config.
If we want to selectively add or remove policies according to the new config, then I think we would have to expose two functions to check if a policy exists and other to remove the policy removed from config.

What do you think about this?

File: lib/gateway/index.js (hunk 8 above)
Comment (ravikp7) [reply lib/gateway/index.js]: `app` parameter is being used to register policy routes

File: lib/gateway/index.js (hunk 7 above)
Comment (ravikp7) [reply lib/gateway/index.js]: sure, will do

File: lib/gateway/index.js (hunk 12 above)
Comment (XVincentX) [reply lib/gateway/index.js]: I'd go in this direction, as long it is not that difficult. The whole idea is to minimize the startup time and the reload time, so it's kind of important to make sure we do not waste time on that.

Comment (ravikp7) [reply lib/gateway/index.js]: I agree with your point. I'm now working on to check and then load or remove core policies as well as plugins.

**Comment by ravikp7** [comment]
@XVincentX Do we also need to remove policies which are removed from the config on hot reload?

**Comment by XVincentX** [comment]
@ravikp7 No; they're on memory already so it does not make sense to unload them. In any case I took care of it and I'm adding an e2e test. See the code in case you're interested!
